anonlink/stats.py

```diff
@@ -0,0 +1,219 @@
+            except StopIteration:
+                return num_matches, num_nonmatches, thresholds
+        if rec_i0 not in matched0 and rec_i1 not in matched1:
+            matched0.add(rec_i0)
+            matched1.add(rec_i1)
+            num_matches[i] += 1
+        else:
+            num_nonmatches[i] += 1
+    return num_matches, num_nonmatches, thresholds
+
+
+def matches_nonmatches_hist(candidate_pairs, bins=100):
+    """Compute a histogram of possible matches and definite nonmatches.
+
+    This function is experimental and subject to change without warning.
+
+    We use the greedy solver to split the candidate pairs into possible
+    matches and definite nonmatches. A possible match may or may not be
+    accepted as a pair depending on the threshold chosen. A definite
+    nonmatch will never be accepted, since one record in this pair has
+    a more promising match with another record. We then make a histogram
+    of possible matches and definite nonmatches.
+
+    :param candidate_pairs: The candidate pairs.
+    :param bins: An integer determining the number of bins to use.
+        Default 100.
+
+    :return: 3-tuple of (1) values of the histogram of the matches as an
+        array of length bins, (2) values of the histogram of the
+        nonmatches as an array of length bins, and (3) the edges of the
+        bins as an array length bins + 1.
+    """
+    semiopen_hist = _semiopen_hist_matches_nonmatches(candidate_pairs, bins)
+    num_matches, num_nonmatches, thresholds = semiopen_hist
+    # Merge last and second last bins for consistency with np.histogram
+    num_matches[-2] += num_matches[-1]
+    num_matches = _np.array(num_matches[:-1])
+    num_nonmatches[-2] += num_nonmatches[-1]
+    num_nonmatches = _np.array(num_nonmatches[:-1])
+    return num_matches, num_nonmatches, thresholds
+
+
+def cumul_number_matches_vs_threshold(candidate_pairs, steps=100):
+    """Compute the number of matches for each threshold.
+
+    This function is experimental and subject to change without warning.
+
+    We use the 2-party greedy solver to calculate the number of matches
+    that would be returned if the candidate_pairs were found using a
+    particular threshold. This function requires only a single pass of
+    the data, so it is faster than simply running the greedy solver
+    multiple times.
+
+    :param candidate_pairs: The candidate pairs.
+    :param steps: An integer determining the number of threshold steps
+        to use. Default 100.
+
+    :return: 2-tuple of (1) the number of matches for the threshold as
+        an array length steps + 1, and (2) the thresholds as an array
+        length steps + 1.
+    """
+    num_matches, _, thresholds = _semiopen_hist_matches_nonmatches(
+        candidate_pairs, steps)
+    num_matches_rev = num_matches[::-1]
+    _np.cumsum(num_matches_rev, out=num_matches_rev)
+    return num_matches, thresholds
+
+
+def plot_similarities_hist(candidate_pairs, bins=100):
```

Comment:
Suspect it would be better to return a `matplotlib.axes.Axes` instead of calling `pyplot.show`. 

I'd probably vote for not including the matplotlib plotting in the library at all though - and move this into an example.

Reply:
Yes, I like the example idea.

I wanted to include the plotting because I had put a bunch of thought into doing it well, and I didn’t want this effort to get lost. But putting it in an example works as well!